### PR TITLE
Merge WCTE to WCSim_v1.11.0

### PR DESCRIPTION
**Pull request by kmtsui**
Merge WCTE/WCSim into the latest WCSim/WCSim

~~Still need to modify the NoReplica construction method to include hybrid config and OD.~~

### CODE REVIEW

File: WCSim.mac [hunk 1]
```diff
@@ -76,7 +76,23 @@
 #/WCSim/WCgeom SuperK_20inchBandL_14perCent # Note: the actual coverage is 13.51%
 #/WCSim/Construct
 
-
+#Use mPMTs settings (uncomment/delete the above)
+#/WCSim/WCgeom nuPRISM_mPMT
+#/WCSim/WCgeom nuPRISMBeamTest_mPMT
+#/WCSim/WCgeom nuPRISMShort_mPMT
```
Comment (tdealtry) [inline WCSim.mac]: Can you elaborate a little on this? Something like

```suggestion
#Use IWCD or WCTE geometry (uncomment/delete the above)
# Use the (old or new?) IWCD geometry
#/WCSim/WCgeom nuPRISM_mPMT
# Use the WCTE geometry
#/WCSim/WCgeom nuPRISMBeamTest_mPMT
# Use the (old or new?) IWCD geometry
#/WCSim/WCgeom nuPRISMShort_mPMT
```

File: WCSim.mac [hunk 2]
```diff
@@ -76,7 +76,23 @@
...
+# True --> default replica method to place PMTs. False --> new placement method
```
Comment (tdealtry) [inline WCSim.mac]: Can you elaborate on new vs old?

File: mPMT-configfiles/mPMTconfig_Position_WCTE.txt [hunk 3]
```diff
@@ -0,0 +1,111 @@
+PMT position table for custom PMT placement. The direction is just for referenec by not used.
```
Comment (tdealtry) [inline mPMT-configfiles/mPMTconfig_Position_WCTE.txt]: You've got a typo here, and a grammar mistake. I think this is what you mean
```suggestion
PMT position table for custom PMT placement. The direction is just for reference and is not used.
```

File: macros/nuPRISMBeamTest.mac [hunk 4]
```diff
@@ -0,0 +1,152 @@
+# nuPRISMBeamTest WCTE macro with no visualization
```
Comment (tdealtry) [inline macros/nuPRISMBeamTest.mac]: Naive question, but why not call this WCTE.mac?

File: macros/tuning_parameters.mac [hunk 5]
```diff
@@ -12,6 +12,13 @@
 /WCSim/tuning/tvspacing 100
 
 
+# Paramater to set photocathode model
+# Set to "0" to use default model, "1" or "2" to use new models
```
Comment (tdealtry) [inline macros/tuning_parameters.mac:16]: What are the new models? Please elaborate. References are also useful

File: macros/tuning_parameters.mac [hunk 6]
```diff
@@ -12,6 +12,13 @@
...
+/WCSim/tuning/pmtsurftype 0
+# Choice of photocathode optical parameters.
+# 0 = (Seems to be) SK. 1 = KCsCb, 2 = RbCsCb from Model2's paper
```
Comment (tdealtry) [inline macros/tuning_parameters.mac]: I'm guessing that these are found in the same paper as **photocathode** model 2?
```suggestion
# 0 = (Seems to be) SK. 1 = KCsCb, 2 = RbCsCb [1 and 2 from from pmtsurftype Model2's paper]
```

File: src/WCSimVisManager.cc [hunk 7]
```diff
@@ -63,6 +63,8 @@
 #include "G4VRML2.hh"
 #endif
 
+#include "G4Version.hh"
+
```
Comment (tdealtry) [inline src/WCSimVisManager.cc]: Already included
```suggestion
```

File: bunny.stl
Comment (tdealtry) [inline bunny.stl:1]: Can you move this into `$WCSIMDIR/macros` or `$WCSIMDIR/data`, just to keep the root dir cleaner?

This also leads to a change in `WCSimConstructCylinder.cc`, updating the (dummy hardcoded) path to this file, in the places where it's commented out.

**Review by tdealtry — CHANGES_REQUESTED** [review]
Thanks for this @kmtsui 

I've had a quick look at the config files and have a few suggestions to make what options mean clearer

My standard validation for HK FD/IWCD/SK is running - will report the results later today

**Comment by tdealtry** [comment]
The current status of my validation between debb4f6 and v1.11.0 is
- nPMTs agree in all geometries (SK/IWCD/HK FD hybrid/HK FD complete). PMT positions from `geofile.txt` also agree
- Number of killed tracks (`GeomNav1002`) consistent (i.e. 0 or 1 per 1000 events in HK/IWCD geometries, up to 12000 per 1000 events in SK)
- `nuPRISMShort_mPMT` geometry has perfect agreement
- SK geometry has differences order 1% in nhits & ndigits
- HK hybrid geometry has differences order 1% in nhits & ndigits (for both PMT types)
- HK complete geometry has 
  - differences order 1% in nhits & ndigits (for all PMT types) for events starting in the ID
  - differences of ~7% for 20"/mPMT nhits & ndigits for events starting in the barrel OD, punching into the ID. OD hits at ~1% level.
  - Weirdly, perfect agreement for all PMT types for events starting in the barrel OD, directed upwards.

Before merging, I would like to understand why there is perfect agreement in IWCD, but not in SK/HK (other than the one for HK OD)

**Comment by kmtsui** [comment]
I prioritized the change of cap/blacksheet geometry to WCTE, let me check this part again

Updated: resolved some difference, can't see why differences happen. Can you run the test again?

**Comment by kmtsui** [comment]
I will need compatibility with ROOT5. Do you mind I push back the old GNUmakefile (with some additional flags) to make it compile?

**Comment by tdealtry** [comment]
Yeah go ahead & push the Makefile back

The important thing is to keep the README accurate with the recommended (and fully supported) setup i.e. cmake, root6 & Geant4.10.3.3

**Comment by kmtsui** [comment]
Ok it seems that the [G4LogicalSkinSurface of border ring blacksheet](https://github.com/WCSim/WCSim/pull/353/commits/aedb7cdb20dc009cc78479d796c09ffe2c52792a) is important to reduce "the photon is killed" warning. This could be the reason for the SK/HK differences

**Comment by tdealtry** [comment]
I've rerun the standard validation on 9fd4a75 and still see the same
- PMT positions identical in all geometries
- Number of killed tracks (GeomNav1002) consistent
- nuPRISMShort_mPMT (all physics) and HK FD complete (only starting in OD barrel, aimed upwards) have perfect agreement
- Other geometries/physics have differences
  - 3-8% difference for events starting in HK FD complete OD barrel, pointing into ID
  - ~1% difference for everything else

I'd still like to understand why there's perfect agreement for some things and not others. Given what the differences are, I presume there is a change to the SK/HK FD (ID part) geometries, but not the IWCD/HK FD (OD part) geometries

**Comment by kmtsui** [comment]
I suspect that is due ot the [G4LogicalSkinSurface of border ring blacksheet](https://github.com/WCSim/WCSim/pull/353/commits/aedb7cdb20dc009cc78479d796c09ffe2c52792a)

Can you try running validation on [224cc4c](https://github.com/WCSim/WCSim/pull/353/commits/224cc4c0bac670f2cc11243373c3a495b1e4176c) instead? This one should give the same G4LogicalSkinSurface setting as the standard WCSim. Or I can push another commit that reverts the setup.

**Comment by tdealtry** [comment]
The CI is working now, mostly fine. But unfortunately I can't see how to get your branch using the new CI file myself. Can you do something like a `git pull && git push` (I think only changes are to `.github/workflows`, so there shouldn't be any conflicts)

**Comment by kmtsui** [comment]
Just did a pull and push. The `TEST_SK_5MeVem_centre_xdir` plots are generated but not the others. Something wrong?

**Comment by tdealtry** [comment]
Think you need to pull again actually. In the build job I was originally doing a quick 5MeV SK test with ~10 events... which failed...

Why did it fail? Simply because you have a different number of stuck track warnings. Your number is much improved (i.e. it is 0)

Anyway I've changed the CI logic such that the build job doesn't run any physics checks, it just sets up the webpage. And I moved the 5MeV SK test into the physics matrix, such that all jobs are guaranteed to run and report their status (assuming the build job completes)

**Comment by kmtsui** [comment]
Correspond to issues https://github.com/WCSim/WCSim/issues/348, https://github.com/WCSim/WCSim/issues/349, https://github.com/WCSim/WCSim/issues/350, https://github.com/WCSim/WCSim/issues/351

**Comment by kmtsui** [comment]
Rotate half a barrel for WCTE from https://github.com/WCTE/WCSim/pull/12

File: WCSim.mac [hunk 8]
```diff
@@ -83,6 +83,29 @@
 #/WCSim/WCgeom SuperK_20inchBandL_14perCent # Note: the actual coverage is 13.51%
 #/WCSim/Construct
 
+#Use IWCD or WCTE geometry (uncomment/delete the above)
+# Use the IWCD geometry
+#/WCSim/WCgeom nuPRISM_mPMT
+# Use the short version of NuPRISM with 6 m tall ID
+#/WCSim/WCgeom nuPRISMShort_mPMT
```
Comment (tdealtry) [inline WCSim.mac]:
```suggestion
#Use WCTE geometry (uncomment/delete the above)
# Use the short version of NuPRISM with 6 m tall ID
#/WCSim/WCgeom nuPRISMShort_mPMT
```
The IWCD options are already in this file above

**Review by tdealtry — CHANGES_REQUESTED** [review]
This is a lot of good work, and it's nice to see that this isn't changing the physics output of things for HK/IWCD. Thanks @kmtsui and the rest of WCTE

I've got one suggestion inline, and a couple of general requests
- In README.md, explain that Makefile is specifically only for use with ROOT5 and is not supported otherwise. All output that needs to be trusted must use cmake
- You’ve added a lot of things to the messenger. It’d be good to document them in WCSim.mac. Here’s a non-exhaustive list
  - gamma-conversion PrimaryGeneratorAction
  - replica, half tower rotation, etc. geometry building
  - A load of optical physics things
  - (Thank you for doing it already for the tuning options in the relevant .mac file)

**Comment by kmtsui** [comment]
More documentations are added. Please check if you want more details in certain aspects.

**Review by tdealtry — APPROVED** [review]

**Pull request merged**